### PR TITLE
[SPARK-19229] [SQL] Disallow Creating Hive Source Tables when Hive Support is Not Enabled

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?
It is weird to create Hive source tables when using InMemoryCatalog. We are unable to operate it. This PR is to block users to create Hive source tables.

### How was this patch tested?
Fixed the test cases

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -370,22 +370,6 @@ trait CheckAnalysis extends PredicateHelper {
...
-
-          // TODO: We need to consolidate this kind of checks for InsertIntoTable
-          // with the rule of PreWriteCheck defined in extendedCheckRules.
-          case InsertIntoTable(s: SimpleCatalogRelation, _, _, _, _) =>
-            failAnalysis(
-              s"""
-                 |Hive support is required to insert into the following tables:
-                 |${s.catalogTable.identifier}
-               """.stripMargin)
-
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: This is not necessary when we blocking users to create Hive source tables. When the table is data source tables, `SimpleCatalogRelation` will be replaced by `LogicalRelation` in the rule `FindDataSourceTable`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 2]
```diff
@@ -758,15 +758,17 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
...
     testSetSerde(isDatasourceTable = false)
   }
 
   test("alter table: set serde (datasource table)") {
     testSetSerde(isDatasourceTable = true)
   }
 
-  test("alter table: set serde partition") {
+  // TODO: move this test to HiveDDLSuite.scala
+  ignore("alter table: set serde partition") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: These two test cases should be moved to HiveDDLSuite. However, it is ugly to copy the codes. Thus, my next plan is to improve the DDLSuite and HiveDDLSuite by creating an abstract class.

**Comment by gatorsmile** [comment]
The following test cases failed:
- change-column.sql
- describe.sql
- show-tables.sql
- show_columns.sql

All these test cases are creating hive serde tables. We might need to create another SQLQueryTestSuite in the hive package.

**Comment by gatorsmile** [comment]
cc @cloud-fan @yhuai @hvanhovell Any comment?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 3]
```diff
@@ -1692,20 +1678,27 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
...
-    (1 to 10).map { i => (i, i) }.toDF("a", "b").createTempView("my_temp_tab")
-    sql(s"CREATE EXTERNAL TABLE my_ext_tab LOCATION '$path'")
-    sql(s"CREATE VIEW my_view AS SELECT 1")
-    intercept[NoSuchTableException] {
-      sql("TRUNCATE TABLE my_temp_tab")
+    withTempPath { tempDir =>
+      withTable("my_ext_tab") {
+        (("a", "b") :: Nil).toDF().write.parquet(tempDir.getCanonicalPath)
+        (1 to 10).map { i => (i, i) }.toDF("a", "b").createTempView("my_temp_tab")
+        sql(s"CREATE TABLE my_ext_tab using parquet LOCATION '$tempDir'")
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: Ah, this is why you asked me https://github.com/apache/spark/pull/16586#discussion_r96142347. I just ran a test for this to help.

```
 - truncate table - external table, temporary table, view (not allowed) *** FAILED *** (188 milliseconds)
   org.apache.spark.sql.AnalysisException: Path does not exist: file:/C:projectsspark	arget	mpspark-9e70280d-56dc-4063-8f40-8e62fec18394;
   at org.apache.spark.sql.execution.datasources.DataSource$$anonfun$14.apply(DataSource.scala:382)
   at org.apache.spark.sql.execution.datasources.DataSource$$anonfun$14.apply(DataSource.scala:370)
   at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
   at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
```

 Maybe, it'd be okay to just use `toURI` if this test is not supposed to test Windows path..

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala [hunk 4]
```diff
@@ -186,6 +186,11 @@ class InMemoryCatalog(
     val db = tableDefinition.identifier.database.get
     requireDbExists(db)
     val table = tableDefinition.identifier.table
+    if (tableDefinition.provider.isDefined && tableDefinition.provider.get.toLowerCase == "hive") {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: shall we put this in `HiveOnlyCheck`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala [hunk 5]
```diff
@@ -278,7 +277,7 @@ abstract class ExternalCatalogSuite extends SparkFunSuite with BeforeAndAfterEac
       schema = new StructType()
         .add("HelLo", "int", nullable = false)
         .add("WoRLd", "int", nullable = true),
-      provider = Some("hive"),
+      provider = Some("parquet"),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogSuite.scala]: shall we also use `defaultProvider` here?

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Pull request closed without merging**